aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -414,26 +450,39 @@ Tensor& add_out_dense_sparse_cpu(Tensor& r, const Tensor& dense, const SparseTen
   int64_t nDim = dense.dim();
   int64_t nDimI = sparse.sparse_dim();
 
-  if (!is_same_tensor(r, dense)) r.copy_(dense);
-  if (sparse._nnz() == 0) return r;
+  if (sparse._nnz() == 0) {
+    if (!is_same_tensor(r, dense)) r.copy_(dense);
+    return r;
+  }
+
+  Tensor values_buffer = maybe_promoted_tensor(values, commonDtype);
```

Comment:
Is it possible to not worry about promoting `values` and just promote the dense and spares tensors being operated on before performing operations? Something like

```
add_out_dense_sparse_cpu(r, dense, sparse) {
   commonDtype = result_type(dense, sparse)
   Tensor dense_wrapper = dense.to(commonDtype)
   Tensor sparse_wrapper = sparse.to(commonDtype)
    // use r, dense_wrapper, and sparse_wrapper in all following operations
}
```

Reply:
It makes more sense if adding the additional layer of indirection suggested above. I'm not sure it actually saves much though to wrap `dense`/`sparse` vs wrapping `dense` and `sparse._values()`. `r` still needs special handling either way since the promoted-result needs to be converted back to the dtype of the result object for in-place ops.